src/ir_passes/bundle_elim.rs

```diff
@@ -0,0 +1,233 @@
+        // delete the original port
+        comp.delete(pidx);
+        // delete the corresponding parameter
+        comp.delete(idx);
+        ports
+    }
+
+    /// Compiles the signature of a component and adds the new ports to the context mapping.
+    fn sig(&self, comp: &mut Component) -> HashMap<PortIdx, Vec<PortIdx>> {
+        // need to collect here because of ownership issues.
+        comp.ports()
+            .idx_iter()
+            .filter(|p| comp.get(*p).is_sig())
+            .collect_vec() // collect here to avoid ownership issues
+            .into_iter()
+            .map(|idx| (idx, self.port(idx, comp)))
+            .collect()
+    }
+
+    /// Compiles the ports defined by an invocation and adds the new ports to the context mapping.
+    /// Mutates the invocation in place, redefining its defined ports.
+    fn inv(
+        &self,
+        idx: InvIdx,
+        comp: &mut Component,
+    ) -> HashMap<PortIdx, Vec<PortIdx>> {
+        let Invoke { ports, .. } = comp.get_mut(idx);
+        let mappings = ports.drain(..).collect_vec();
+        let mappings = mappings
+            .into_iter()
+            .map(|p| (p, self.port(p, comp)))
+            .collect_vec();
+
+        comp.get_mut(idx)
+            .ports
+            .extend(mappings.iter().flat_map(|(_, v)| v));
+
+        mappings.into_iter().collect()
+    }
+
+    fn connect(
+        &self,
+        connect: &Connect,
+        cidx: CompIdx,
+        ctx: &mut Context,
+    ) -> Vec<Command> {
+        let comp = ctx.get_mut(cidx);
+        let mapping = self.context.get(cidx);
+
+        let Connect { src, dst, info } = connect;
+
+        let start = src.start.as_concrete(comp).unwrap() as usize;
+        let end = src.end.as_concrete(comp).unwrap() as usize;
+        let src = mapping.get(&src.port).unwrap()[start..end].to_vec();
+
+        let start = dst.start.as_concrete(comp).unwrap() as usize;
+        let end = dst.end.as_concrete(comp).unwrap() as usize;
+        let dst = mapping.get(&dst.port).unwrap()[start..end].to_vec();
+
+        assert!(
+            src.len() == dst.len(),
+            "Mismatched access lengths for connect `{}`",
+            connect
+        );
+
+        src.into_iter()
+            .zip(dst.into_iter())
+            .map(|(src, dst)| {
+                (Access::port(src, comp), Access::port(dst, comp))
+            })
+            .map(|(src, dst)| {
+                Command::Connect(Connect {
+                    src,
+                    dst,
+                    info: *info,
+                })
+            })
+            .collect()
+    }
+
+    /// Compiles the body of a component and replaces all ports with their expanded versions.
+    fn comp(&mut self, cidx: CompIdx, ctx: &mut Context) {
+        let comp = ctx.get_mut(cidx);
+        // compile invocations
+        for idx in comp.invocations().idx_iter() {
+            let pl = self.inv(idx, comp);
+            self.context.get_mut(cidx).extend(pl);
```

Comment:
Should we be deleting old `invoke`s here? The `invoke` call does not. Alternatively, we can write a "dead code elimination" pass in that can just remove all unused things (ports, invokes, instances) from a component.

Reply:
This mutates the invoke, and doesn't generate new ones, so nothing is created/deleted.